_posts/2025-04-23-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,202 @@
+---
+layout: post
+title: 'Getting ready for secure MCP with Quarkus MCP Server'
+date: 2025-04-23
+tags: ai mcp security
+synopsis: 'Demonstrate secure Quarkus MCP SSE server access in the dev mode'
+author: sberyozkin
+---
+:imagesdir: /assets/images/posts/secure_mcp_sse_server
+
+== Introduction
+
+https://modelcontextprotocol.io/specification/2025-03-26[The latest version of the Model Context Protocol (MCP) specification] introduces an https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization[authorization] flow for HTTP based MCP transports.
+
+MCP authorization will enable MCP SSE and Streamable HTTP clients to access MCP servers securely and is being very actively discussed and evaluated right now. This https://github.com/modelcontextprotocol/modelcontextprotocol/pull/284[MCP authorization PR] is one of the focal points for the MCP authorization discussions.
+
+While it will take a bit of time for the MCP authorization specificaion to finalize and be widely supported, one thing you can be certain about is that secure MCP servers that wish to interoperate with MCP authorization aware MCP clients will be required to accept and verify bearer access tokens sent by these clients on behalf of their users.
+
+The https://github.com/quarkiverse/quarkus-mcp-server[Quarkus MCP Server project] is ready for you to start experimenting with secure MCP SSE servers.
```

Comment:
We would need a compatible client to build a story.

Reply:
just to clarify - I thought quarkus-mcp-server did *not* yet implement 2025-03-26 spec.

are we just saying here that the oauth in the new spec is possible because it literally just is oauth which is flexible enough in quarkus to apply here or?

Reply:
now i reread it - this is explaining how to secure mcp servers in general - and it does NOT require the latest version of the spec nor a compatible client (besides it should be able to send a header token).

I think that should be made clearer in that introduction that this is something that works for MCP *today*.

Reply:
@cescoffier @maxandersen Thanks, yes, this post is about showing how, given a token, users can access Quarkus MCP server from either Dev UI or curl, given that none of the existing MCP clients can actually use OAuth2 to authenticate users and forward tokens automatically.

>I think that should be made clearer in that introduction that this is something that works for MCP today.

I'm not sure about it, I believe we don't have MCP clients that can automatically consume MCP SSE server which responds 401 if no token is sent

Reply:
@maxandersen Sure, this is just not the MCP authorization spec but as agreed we will cover it